### PR TITLE
Preliminary fleshing out of writers for binary/string types

### PR DESCRIPTION
**Pull request by maxburke**
This doesn't have tests yet; I'll work on those next. Figured I'd start stitching this together to see if I'm on the right path.

### CODE REVIEW

File: rust/parquet/src/arrow/arrow_writer.rs [hunk 1]
```diff
@@ -144,12 +145,17 @@ fn write_leaves(
...
+        | ArrowDataType::LargeBinary
+        | ArrowDataType::Binary
+        | ArrowDataType::Utf8
+        | ArrowDataType::LargeUtf8 => {
             let mut col_writer = get_col_writer(&mut row_group_writer)?;
             write_leaf(
                 &mut col_writer,
                 array,
                 levels.pop().expect("Levels exhausted"),
+                array.data_type(),
```
Comment (nevi-me) [inline rust/parquet/src/arrow/arrow_writer.rs]: you need not pass the data type as you can get it from the array

File: rust/parquet/src/arrow/arrow_writer.rs [hunk 2]
```diff
@@ -430,6 +472,28 @@ fn get_primitive_def_levels(
     primitive_def_levels
 }
 
+macro_rules! def_get_binary_array_fn {
+    ($name:ident, $ty:ty) => {
+        fn $name(array: &$ty) -> Vec<ByteArray> {
+            let mut values = Vec::with_capacity(array.len() - array.null_count());
+            for i in 0..array.len() {
+                if array.is_valid(i) {
+                    let buffer = BufferPtr::new(array.value(i).as_bytes().into());
```
Comment (nevi-me) [inline rust/parquet/src/arrow/arrow_writer.rs]: This looks longer than `ByteArray::from(array.value(i).to_vec())`. Given that the difference btwn BinaryArray and StringArray is that the latter has valid utf8 bytes, when you only want the bytes, it's convenient to reinterpret StringArray as BinaryArray using `BinaryArray::from(array.data())`.

**Review by nevi-me — COMMENTED** [review]
I worked on the binary writer yesterday, but hadn't pushed my changes because I'm still struggling with indexing into deeply nested primitives.

My approach looks like below:

```rust
fn get_byte_array_slice(
    array: &arrow_array::ArrayRef,
    def_levels: &[i16],
    max_def_level: i16,
    // _offsets: &[usize],
) -> Vec<ByteArray> {
    let slice_len = def_levels.iter().filter(|v| **v == max_def_level).count();
    let mut values = Vec::with_capacity(slice_len);

    match array.data_type() {
        ArrowDataType::Binary | ArrowDataType::Utf8 => {
            let binary_array = arrow_array::BinaryArray::from(array.data());
            // TODO: only take values where def_level = max_def_level
            for i in 0..array.len() {
                if array.is_valid(i) {
                    values.push(ByteArray::from(binary_array.value(i).to_vec()))
                }
            }
        }
        ArrowDataType::LargeBinary | ArrowDataType::LargeUtf8 => {
            let large_binary_array = arrow_array::LargeBinaryArray::from(array.data());
            // TODO: only take values where def_level = max_def_level
            for i in 0..array.len() {
                if array.is_valid(i) {
                    values.push(ByteArray::from(large_binary_array.value(i).to_vec()))
                }
            }
        }
        _ => unreachable!("Function should only be called with supported datatypes")
    };

    values
}
```

The `def_levels` and `max_def_levels` aren't used yet because I'm still bikeshedding a way of indexing into array values that only have the definition levels that should be written into.

File: rust/parquet/src/arrow/arrow_writer.rs (hunk 2 above)
Comment (maxburke) [reply rust/parquet/src/arrow/arrow_writer.rs]: :+1:

File: rust/parquet/src/arrow/arrow_writer.rs (hunk 1 above)
Comment (maxburke) [reply rust/parquet/src/arrow/arrow_writer.rs]: :+1:

**Comment by maxburke** [comment]
Added a few tests for writing / reading the binary and string types.

**Review by nevi-me — APPROVED** [review]

**Pull request merged**